### ISSUE TITLE
Undo does not revert the processed region

### ISSUE THREAD

**Issue by JasonGross** [ProofGeneral/PG#39]
If I `C-/`, and this deletes a word from the already processed region, the region is not reverted.

**Comment by erikmd** [comment]
Normally, this kind of issue should be fixed since PG 4.0.
Can you check what is your value of variable `proof-strict-read-only` ?
The default value is `'retract`, while a `nil` value actually tells PG not to revert the processed region.

**Comment by cpitclaudel** [comment]
Can you make a small repro? Things work fine for me.

**Comment by JasonGross** [comment]
Hmmmm, maybe I was using an old version of PG last time it happened.  I did hit another bug, while messing around with this, but I'll open a new issue for that.

**Comment by JasonGross** [comment]
Oh, no, I managed to hit it again.  Let me make an example.

**Comment by JasonGross** [comment]
It seems that an Undo of a paste is exempt from this rule.  Put the following in your buffer:
```coq
Definition False := True.

Theorem bad : False.
Proof.
  constructor.
Qed.
```
Now type `M-< C-k C-y M-> C-c C-RET C-/` (cursor to beginning, cut, paste, cursor to end, evaluate document, undo), and see ProofGeneral accepting a document which claims to prove `False`.

**Comment by JasonGross** [comment]
Note that if I then `C-/` again (to re-paste), the document is in an inconsistent state; trying to evaluate the first line tells me `Error: False already exists.`

**Comment by JasonGross** [comment]
(`proof-strict-read-only` is `retract`, and I'm using PG Version 4.4pre)

**Comment by erikmd** [comment]
Hi Jason,
Thanks for your MWE ! I was able to reproduce the issue. I'll look into this and keep you informed.
Cheers,
Erik

**Comment by Matafou** [comment]
Hi,

I don't understand, C-/ is exactly made to bypass the locked region
protection. This is not a bug to me, but a feature. If you want PG to
backtrack when undoing in locked region, just use the usual undo command
(C-x u or C-_).

BUT: I notice that on emacs 25.1 it does backtrack even when using C-/.
Which I do consider a bug.

P.

2016-07-04 12:18 GMT+02:00 Erik Martin-Dorel <notifications@github.com>:
>
> Hi Jason,
> Thanks for your MWE ! I was able to reproduce the issue. I'll look into
this and keep you informed.

**Comment by JasonGross** [comment]
I was not aware there was any other undo command (and both of those take an extra key stroke).  I think the default should be that newcomers can't shoot themselves in the foot, and if you want `C-/` to not revert the locked region, then there could be an option to enable that.

**Comment by cpitclaudel** [comment]
Hi Pierre and Jason,

Jason: This must be a slow day for me: I can't repro your problem, regardless of the key I use.

Pierre: These keys seem to be bound to the same command?

```
C-/ runs the command pg-protected-undo (found in coq-mode-map), which is an
interactive Lisp function in ‘pg-user.el’.

It is bound to C-_, <undo>, C-/, C-x u, <menu-bar> <edit> <undo>.

(pg-protected-undo &optional ARG)

As ‘undo’ but avoids breaking the locked region.

It performs each of the desired undos checking that these operations will
not affect the locked region, obeying ‘proof-strict-read-only’ if required.
If strict read only behaviour is enforced, the user is queried whether to
retract before the undo is allowed.  If automatic retraction is enabled,
the retract and undo will go ahead without querying the user.

Moreover, undo/redo is always allowed in comments located in the locked region.

[back]
```

**Comment by Matafou** [comment]
I agree.

There is a need for a "one shot" shortcut IMHO, but the choice of C-/ is bad indeed. I wonder what was the motivation...

I will remove the shortcut and this one
C-c C-a C-/
ok?

It remains that this feature is broken with recent emacsen.

P.

**Comment by cpitclaudel** [comment]
Hmm, looks like we posted at the same time. I'm fine with `C-c C-a C-/`... but I can't see where in the current PG such a thing is defined, since for me `C-/` is bound to the same as `C-_`

**Comment by Matafou** [comment]
My mistake, I did not read the documentation well enough. C-/ is supposed to be safe. this is a real bug, but it seems to be emacs-24 specific.

P.

**Comment by Matafou** [comment]
They are bound to the same. I misread the code. undo is globally rebound to
this protected version, which in my opinion is good.

I do reproduce but only with emacs-24. Jason what is your version of emacs
please?

P.

2016-07-04 15:30 GMT+02:00 Clément Pit--Claudel <notifications@github.com>:

**Comment by erikmd** [comment]
Hi,

Le lundi 4 juillet 2016 à 04:47 -0700, Pierre Courtieu a écrit :
> I don't understand, C-/ is exactly made to bypass the locked region
> protection.

The issue mentioned by Jason is also reproducible with [C-_] for me
(Emacs 24.3.1), and after a big of investigating it is due the
`next-undo-elt` function that I had contributed for refactoring the
`pg-protected-undo` function in PG 4.0.

Namely, this function indeed returns the latest undo element in
the current buffer, but forgets to skip the irrelevant undo elements
that may be in front of the list after a yank, so that
`(next-undo-elt nil)` wrongly returns
`(nil rear-nonsticky nil 25 . 26)` instead of 
`(1 . 26)`

For the recard, here is the content of the variable `buffer-undo-list`
with Jason's example:
```
Value: (nil
 (nil rear-nonsticky nil 25 . 26)
 (nil fontified nil 25 . 26)
 (nil fontified nil 21 . 25)
 (nil fontified nil 17 . 21)
 (nil fontified nil 12 . 17)
 (nil fontified nil 11 . 12)
 (nil fontified nil 1 . 11)
 (1 . 26)
 nil
 (#("Definition False := True." 0 10
    (fontified t face font-lock-keyword-face)
    10 11
    (fontified t)
    11 16
    (fontified t face font-lock-type-face)
    16 20
    (fontified t)
    20 24
    (fontified t face font-lock-type-face)
    24 25
    (fontified t))
  . 1)
 (t 22394 26246 208417 173000))
```

I'm working on a fix and I'll submit a PR for that.
Best,
Erik

**Comment by erikmd** [comment]
BTW, the syntax of the undo elements is described here:
https://www.gnu.org/software/emacs/manual/html_node/elisp/Undo.html

**Comment by cpitclaudel** [comment]
Neat. Still, the bug doesn't seem to exist in Emacs 25; I wonder why... (POPL deadline means I don't have time to look at code atm, sorry :/)

**Comment by Matafou** [comment]
The function described by Erik has been fixed in emacs-25?
P.

2016-07-04 15:45 GMT+02:00 Clément Pit--Claudel <notifications@github.com>:

**Comment by cpitclaudel** [comment]
Isn't the issue in a PG function?

**Comment by erikmd** [comment]
Yes, it is due to the PG function `(next-undo-elt)`. So I don't know why the issue does not pop up in emacs-25 as well.

**Comment by Matafou** [comment]
OK but this function calls emacs builtin functions like buffer-undo-list
and use emacs builtin variables like buffer-undo-list.

P.

2016-07-04 15:55 GMT+02:00 Erik Martin-Dorel <notifications@github.com>:

**Comment by erikmd** [comment]
Hi @Matafou,

Indeed, it appears that the implementation of emacs-25 yields a "simpler" value for `buffer-undo-list` with @JasonGross's MWE, which explains the absence of issue with emacs-25...

To be more precise, I ran:
```sh
git clone --depth=1 'git://git.savannah.gnu.org/emacs.git' --branch emacs-25 /tmp/emacs
cd /tmp/emacs/
./autogen.sh 
./configure --prefix="$HOME" --without-all --with-x-toolkit=no --without-x --with-xml2 
make -j3 install
emacs -nw
```
and replaying Jason's MWE led to the following value of `buffer-undo-list`:
```
Value: (nil
 (1 . 26)
 nil
 (#("Definition False := True." 0 10
    (fontified t face font-lock-keyword-face)
    10 11
    (fontified t)
    11 16
    (fontified t face font-lock-type-face)
    16 20
    (fontified t)
    20 24
    (fontified t face font-lock-type-face)
    24 25
    (fontified t))
  . 1)
 (#<marker in no buffer> . -25)
 (t 22394 32584 944874 820000))
```
EMD

**Comment by JasonGross** [comment]
Thanks for tracking this down!  Indeed, I'm using emacs 24.something on Windows.

**Comment by erikmd** [comment]
Hi @Matafou and @JasonGross, as promised I submitted a fix in PR #87; since you both reproduced this issue with emacs24, could one of you check if #87 also solves the problem for you?
Kind regards.

**Comment by Matafou** [comment]
It solves the problem on my emacs-24. Thanks Erik.
P.
